crates/witgen/src/canoe_witness_provider.rs

```diff
@@ -0,0 +1,55 @@
+use alloy_consensus::Header;
+use alloy_rlp::Decodable;
+use canoe_provider::{CanoeInput, CanoeProvider};
+use core::fmt::Debug;
+use hokulea_proof::eigenda_blob_witness::EigenDABlobWitnessData;
+use kona_preimage::{CommsClient, PreimageKey};
+use kona_proof::{BootInfo, FlushableCache};
+use std::sync::Arc;
+
+/// A helper function to create canoe proof by the provided canoe provider.
+/// The function relies on data stored in the oracle, for l1_head, l1_head_number
+/// and chain_id.
+pub async fn from_boot_info_to_canoe_proof<P, O>(
+    boot_info: &BootInfo,
+    witness: &EigenDABlobWitnessData,
+    oracle: Arc<O>,
+    canoe_provider: P,
+) -> anyhow::Result<Vec<P::Receipt>>
+where
+    P: CanoeProvider,
+    O: CommsClient + FlushableCache + Send + Sync + Debug,
+{
+    let header_rlp = oracle
+        .get(PreimageKey::new_keccak256(*boot_info.l1_head))
+        .await
+        .expect("get l1 header based on l1 head");
```

Comment:
your function returns an anyhow::Result but you use `expect()` everywhere instead of propagating errors. Either return a Result and propagate errors with `?`, or instead use `expect()` everywhere and return a `Vec<P::Receipt>` directly (not wrapped in Result)